modules/geolocation/geolocation.py

```diff
@@ -585,11 +604,25 @@ def run_locator(self, telemetry, coordinates):
         if len(point_pairs) < 4:
             return False, None
 
-        non_collinear_points = self.get_non_collinear_points(point_pairs)
+        # Slice point_pairs to shape (n, 2) for input of get_non_collinear_points
+        points = point_pairs[:,1]
+
+        # Get the 4 non-collinear indexes of the array above
+        indexes = self.get_non_collinear_points(points)
+
         # If insufficient point pairs, exit this run and try again
-        if len(non_collinear_points) < 4:
+        if len(indexes) < 4:
             return False, None
 
+        non_collinear_points = np.empty(shape=(0, 2, 2))
+        # Create a subset of the (n, 2, 2) array above using the array of indexes
+        for i in range(0, len(indexes)):
```

Comment:
Replace loops with just `non_collinear_points = point_pairs[indexes]` to extract the points we want (see [this answer](https://stackoverflow.com/a/3179137)).

As a side note that won't be applied given the above:

`i == j` is a bug and should be `indexes[i] == j` . But if we already know what the indexes are we don't actually need the inner loop, and can just take them with `point_pairs[indexes[i]]` instead. Also, `numpy.append()` and `numpy.concatenate()` is expensive (underneath, a new array is allocated with a bigger size and elements copied over); since we already know the size it would be better to start with an empty array of size 4 instead of 0.